File: 2005/Windows SDK NetFx 3.0/Technologies/DirectoryServices/DirSyncSearch/CS/DirSyncSearch/DirSyncSearch.cs
/*=====================================================================
  File:      DirSyncSearch.cs

  Summary:   This sample shows how directory synchronization search
             works in System.DirectoryServices

---------------------------------------------------------------------
  This file is part of the Microsoft .NET SDK Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

This source code is intended only as a supplement to Microsoft
Development Tools and/or on-line documentation.  See these other
materials for detailed information regarding Microsoft code samples.

THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
PARTICULAR PURPOSE.
=====================================================================*/

// To compile: 
//      csc.exe DirectorySynchronization.cs 
// To run: 
//      DirectorySynchronization.exe 

using System;
using System.Security.Permissions;

using System.DirectoryServices;


[assembly: System.Runtime.InteropServices.ComVisible(false)]
[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]
[assembly: SecurityPermission(SecurityAction.RequestMinimum, Execution = true)]
[assembly: DirectoryServicesPermission(SecurityAction.RequestMinimum)]

namespace Microsoft.Samples.DirectoryServices
{
    public class DirectorySynchronizationSearchSample
    {
        // Change the value of next constant to bind to a particular domain
        // Value of null means current domain
        static string domainADsPath = null;

        public static void Main(string[] args)
        {
            try
            {
                DirectorySynchronizationSearch();

                Console.WriteLine("\r\nApplication Finished Successfully!!!");
            }
            catch (Exception e)
            {
                Console.WriteLine("\r\nUnexpected exception occured:\r\n\t" +
                                e.GetType() + ":" + e.Message);
            }            
        }

        static void DirectorySynchronizationSearch()
        {
            // First add a temporal user to help the sample
            string lDAPPathUser = addTempUser(); 
            
            DirectorySearcher searcher = new DirectorySearcher();

            
            searcher.SearchRoot = new DirectoryEntry(domainADsPath);
            searcher.Filter = "(objectClass=user)";
            // For a directory synchronization search search scope must be subtree
            searcher.SearchScope = SearchScope.Subtree;
            searcher.Sort = new SortOption(
                                "name", SortDirection.Ascending);

            searcher.PropertiesToLoad.AddRange(
                            new string[] { "displayName", "name" });

            searcher.DirectorySynchronization = new DirectorySynchronization();

            // Do the first search - results should be all
            SearchResultCollection results = searcher.FindAll();
            Console.WriteLine("DO THE FIRST SEARCH");
            displayResults(results);
            // Results are disposed explicitly
            results.Dispose();

            // Do the same search - results should be none
            SearchResultCollection newResults = searcher.FindAll();
            Console.WriteLine("\r\n\r\nDO THE SAME SEARCH");
            displayResults(newResults);
            // Results are disposed explicitly
            newResults.Dispose();

            // Modify displayName property value
            modifyUser(lDAPPathUser, "_MODIFIED");

            // Do the same search - results should be modified user
            newResults = searcher.FindAll();
            Console.WriteLine("\r\n\r\nDO THE SAME SEARCH");
            displayResults(newResults);
            // Results are disposed explicitly
            newResults.Dispose();

            // Delete temporary user
            deleteTempUser(lDAPPathUser);
        }

        private static string addTempUser()
        {
            DirectoryEntry entry = new DirectoryEntry(domainADsPath);

            DirectoryEntry newEntry = 
                        entry.Children.Add("CN=TemporaryUser", "user");

            newEntry.Properties["displayName"].Value = "TemporaryUser";
            newEntry.CommitChanges();

            return newEntry.Path;
        }

        private static void deleteTempUser(string lDAPPathUser)
        {
            DirectoryEntry entry = new DirectoryEntry(domainADsPath);

            entry.Children.Remove(new DirectoryEntry(lDAPPathUser));
        }

        private static void modifyUser(string lDAPPAth, string stringToAdd)
        {
            DirectoryEntry entry = new DirectoryEntry(lDAPPAth);

            entry.Properties["displayName"].Value += stringToAdd;
            entry.CommitChanges();
        }

        private static void displayResults(SearchResultCollection results)
        {
            // About to perform search
            Console.Write("Searching...");
            if (results.Count == 0) 
                Console.WriteLine("\r\nThere is no results to display");
            else
            {
                Console.WriteLine("\r\nRESULT COLLECTION INFORMATION");
                Console.WriteLine("Number of results:\t\t" + 
                                        results.Count.ToString());
                Console.WriteLine("Number of properties loaded:\t" + 
                                        results.PropertiesLoaded.Length);
                Console.Write("Properties loaded:\t\t");
                for (int i = 0; i < results.PropertiesLoaded.Length; i++)
                    Console.Write(results.PropertiesLoaded[i] + " ");
                Console.WriteLine("\r\n\r\nRESULTS");

                foreach (SearchResult result in results)
                {
                    Console.WriteLine("Path:\t\t" + result.Path);
                    Console.WriteLine("Display Name:\t" + 
                                        result.Properties["displayName"]);
                    Console.WriteLine("Name:\t\t" + 
                                        result.Properties["name"]);
                }
            }
        }
    } // End class DirectorySynchronizationSearchSample
}

File: 2005/Windows SDK NetFx 3.0/Technologies/DirectoryServices/DirSyncSearch/VB/DirSyncSearch/DirSyncSearch.vb
'=====================================================================
'  File:      DirSyncSearch.vb
'
'  Summary:   This sample shows how directory synchronization search
'             works in System.DirectoryServices
'
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET SDK Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
'
'THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'=====================================================================

Imports System
Imports System.Security.Permissions

Imports System.DirectoryServices


<Assembly: System.Reflection.AssemblyVersion("1.0.0.0")> 
<Assembly: SecurityPermission(SecurityAction.RequestMinimum, Execution:=True)> 
<Assembly: DirectoryServicesPermission(SecurityAction.RequestMinimum)> 


Namespace Microsoft.Samples.DirectoryServices

    Public NotInheritable Class DirectorySynchronizationSearchSample
        ' Change the value of next constant to bind to a particular domain
        ' Value of null means current domain
        Private Shared domainADsPath As String = "LDAP://fabrikam.com"

        Public Overloads Shared Sub Main()
            Try
                DirectorySynchronizationSearch()

                Console.WriteLine()
                Console.WriteLine("Application Finished Successfully!!!")
            Catch e As Exception
                Console.WriteLine()
                Console.WriteLine("Unexpected exception occured:")
                Console.WriteLine(e.GetType().ToString() + ":" + e.Message)
            End Try
        End Sub 'Main

        Private Sub New()
        End Sub

        Shared Sub DirectorySynchronizationSearch()
            ' First add a temporal user to help the sample
            Dim lDAPPathUser As String = addTempUser()

            Dim searcher As New DirectorySearcher()

            searcher.SearchRoot = New DirectoryEntry(domainADsPath)
            searcher.Filter = "(objectClass=user)"
            ' For a directory synchronization search scope must be subtree
            searcher.SearchScope = SearchScope.Subtree
            searcher.Sort = New SortOption("name", _
                            SortDirection.Ascending)

            searcher.PropertiesToLoad.AddRange( _
                            New String() {"displayName", "name"})

            searcher.DirectorySynchronization = New DirectorySynchronization()

            ' Do the first search - results should be all
            Dim results As SearchResultCollection = searcher.FindAll()
            Console.WriteLine("DO THE FIRST SEARCH")
            displayResults(results)
            ' Results are disposed explicitly
            results.Dispose()

            ' Do the same search - results should be none
            Dim newResults As SearchResultCollection = searcher.FindAll()
            Console.WriteLine()
            Console.WriteLine()
            Console.WriteLine("DO THE SAME SEARCH")
            displayResults(newResults)
            ' Results are disposed explicitly
            newResults.Dispose()

            ' Modify displayName property value
            modifyUser(lDAPPathUser, "_MODIFIED")

            ' Do the same search - results should be modified user
            newResults = searcher.FindAll()
            Console.WriteLine()
            Console.WriteLine()
            Console.WriteLine("DO THE SAME SEARCH")
            displayResults(newResults)
            ' Results are disposed explicitly
            newResults.Dispose()

            ' Delete temporary user
            deleteTempUser(lDAPPathUser)
        End Sub 'DirectorySynchronizationSearch


        Private Shared Function addTempUser() As String
            Dim entry As New DirectoryEntry(domainADsPath)

            Dim newEntry As DirectoryEntry = _
                            entry.Children.Add("CN=TemporaryUser", "user")

            newEntry.Properties("displayName").Value = "TemporaryUser"
            newEntry.CommitChanges()

            Return newEntry.Path
        End Function 'addTempUser


        Private Shared Sub deleteTempUser(ByVal lDAPPathUser As String)
            Dim entry As New DirectoryEntry(domainADsPath)

            entry.Children.Remove(New DirectoryEntry(lDAPPathUser))
        End Sub 'deleteTempUser


        Private Shared Sub modifyUser(ByVal lDAPPAth As String, _
                                        ByVal stringToAdd As String)
            Dim entry As New DirectoryEntry(lDAPPAth)

            entry.Properties("displayName").Value += stringToAdd
            entry.CommitChanges()
        End Sub 'modifyUser


        Private Shared Sub displayResults( _
                                ByVal results As SearchResultCollection)
            ' About to perform search
            Console.Write("Searching...")
            If results.Count = 0 Then
                Console.WriteLine()
                Console.WriteLine("There is no results to display")
            Else
                Console.WriteLine()
                Console.WriteLine("RESULT COLLECTION INFORMATION")
                Console.WriteLine(("Number of results:" + _
                                    results.Count.ToString()))
                Console.WriteLine(("Number of properties loaded:" + _
                                    results.PropertiesLoaded.Length.ToString()))
                Console.Write("Properties loaded:")
                Dim i As Integer
                For i = 0 To results.PropertiesLoaded.Length - 1
                    Console.Write((results.PropertiesLoaded(i) + " "))
                Next i
                Console.WriteLine()
                Console.WriteLine()
                Console.WriteLine("RESULTS")

                Dim result As SearchResult
                For Each result In results
                    Console.WriteLine(("Path:" + result.Path))
                    Console.WriteLine(("Display Name:" + _
                                result.Properties("displayName").ToString()))
                    Console.WriteLine(("Name:" + _
                                result.Properties("name").ToString()))
                Next result
            End If
        End Sub 'displayResults
    End Class 'DirectorySynchronizationSearchSample
End Namespace 'Microsoft.Samples.DirectoryServices